[RHCLOUD-18212] Refactor the way DAOs are created

MikelAlejoBR opened the pull request:
The reason behind this change is to decouple the instantiation of the DAOs from the code that needs them, in order to gain more flexibility when switching DAO implementations or when testing. Consider this example:

```go
// dao_file.go
type Dao interface {
	FindByAccountNumber(string accountNumber) (Something, error)
}

type dbDaoImpl struct {}

func (d* dbDaoImpl) FindByAccountNumber(string accountNumber) (Something, error) {
	// call db and return whatever
}

type vaultDaoImpl struct {}

func (d* dbDaoImpl) FindByAccountNumber(string accountNumber) (Something, error) {
	// call Vault and return whatever
}

// my_file_that_uses_the_dao.go
func myFunctionThatNeedsADao(string accountNumber) error {
	dao := dbDaoImpl{} // now this function is coupled to the dbDao, and we
	                   // cannot change to the vaultDao implementation.

	something, err := dao.FindByAccountNumber(&accountNumber)

	// ...

	return nil
}
```
In the above case the implementation is very coupled to the function that uses it, so... what could we do if we wanted to use the `VaultDaoImpl` instead? :thinking: 

```go
// dao_file.go
var GetDao func() Dao // returns the interface!

var getDefaultDao func() Dao {
	if config.VaultIsOnline {
		return &vaultDaoImpl{}
	} else {
		return &dbDaoImpl{}
	}
}

// init will set the function for obtaining the dao to the default one we have
// coded. If you want to use a different one, to return a mock dao for example,
// you only have to set another function to the exported GetDao variable!
func init() {
	GetDao = getDefaultDao
}

// my_file_that_uses_the_dao.go
func myFunctionThatNeedsADao(string accountNumber) error {
	// In this case we don't care which DAO implementation is underlying, since
	// we only care that we will be returned "something".
	dao := GetDao()
	something, err := dao.FindByAccountNumber(&accountNumber)

	// ...

	return nil
}
```

This is one way of doing it. We can override the `GetDao` function whenever we want if we need a different DAO.

---

So, this PR:

- Extracts the "get tenant from context" part of the code that was duplicated in every handler to a helper function.
- Creates a `var GetDao*` function for every DAO, and initializes it with a default DAO provider on the `init()` function, so that every other package has only one way of using DAOs, and so that those packages work against interfaces and not final implementations.
- Makes the `*Impl` structs unexported so that nobody can instantiate a DAO directly.

## Links

[[RHCLOUD-18212]](https://issues.redhat.com/browse/RHCLOUD-18212)

MikelAlejoBR committed:
refactor: helper function to pull the tenant from the echo context

Instead of repeating the same code over the multiple handlers we've got,
this should help centralizing the logic of pulling the tenant from the
request.

MikelAlejoBR committed:
refactor: make all the handlers use the helper function

MikelAlejoBR committed:
refactor: use dao provider function for ApplicationAuthentications

MikelAlejoBR committed:
refactor: use dao provider function for Applications

MikelAlejoBR committed:
refactor: use dao provider function for ApplicationTypes

MikelAlejoBR committed:
refactor: use dao provider function for Authentications

MikelAlejoBR committed:
refactor: use dao provider function for Endpoint

MikelAlejoBR committed:
refactor: use dao provider function for MetaDatas

MikelAlejoBR committed:
refactor: use dao provider function for RhcConnections

MikelAlejoBR committed:
refactor: use dao provider function for Sources

MikelAlejoBR committed:
refactor: make TenantDAO an interface and create a provider function

lindgrenj6 reviewed (approved):
Awesome, thanks for getting this out so quick. This will enforce how we isntantiate db dao's all over the app! I'll have to refactor a bunch in my other PRs haha

MikelAlejoBR committed:
refactor: use dao provider function for SourceTypes

MikelAlejoBR committed:
refactor: make DAOs satisfy the EventModelDao interface

This is required for the Source, Endpoint, Application and
Authentication DAOs in order to be able to use the provider functions in
the "common.go" code.

MikelAlejoBR committed:
refactor: use provider functions in DAO's common code

petracihalova reviewed (approved):
Looks great. I really appreciate the way you split changes into commits 👍🏻 .
